### PR TITLE
Validate arguments for `ps` in `docker top`

### LINKED ISSUE #24357: `docker top` should validate arguments for `ps`
**Output of `docker version`:**

    Client:
     Version:      1.12.0-rc2
     API version:  1.24
     Go version:   go1.6.2
     Git commit:   906eacd-unsupported
     Built:        Fri Jun 17 21:12:28 2016
     OS/Arch:      linux/amd64
     Experimental: true
    
    Server:
     Version:      1.12.0-rc2
     API version:  1.24
     Go version:   go1.6.2
     Git commit:   906eacd-unsupported
     Built:        Fri Jun 17 21:12:28 2016
     OS/Arch:      linux/amd64
     Experimental: true

**Output of `docker info`:**

    Containers: 2
     Running: 2
     Paused: 0
     Stopped: 0
    Images: 1
    Server Version: 1.12.0-rc2
    Storage Driver: overlay
     Backing Filesystem: extfs
    Logging Driver: json-file
    Cgroup Driver: cgroupfs
    Plugins:
     Volume: local
     Network: host bridge null overlay
    Swarm: inactive
    Runtimes: default
    Default Runtime: default
    Security Options: apparmor seccomp
    Kernel Version: 4.4.0-24-generic
    Operating System: Ubuntu 16.04 LTS
    OSType: linux
    Architecture: x86_64
    CPUs: 4
    Total Memory: 15.6 GiB
    Name: suda-PC2
    Docker Root Dir: /var/lib/docker
    Debug Mode (client): false
    Debug Mode (server): false
    Registry: https://index.docker.io/v1/
    WARNING: No swap limit support
    Experimental: true
    Insecure Registries:
     127.0.0.0/8

In Ubuntu 16.04, `/bin/ps` is provided by `procps_3.3.10-4ubuntu2_amd64.deb`.

**Steps to reproduce the issue:**

1. Suppose that a user `u1` has UID=1000.
2. Suppose that a container `c1` has a process of which PID is 1000 (in the host). Actually the PID is unlike to correspond with the UID of `u1`, but attacker can increase the probability by running `docker run` for just forking repeatedly. Attacker can also utilize GID or something else instead of UID.
3. Run `sudo -u u1 -E SECRET=foobar cat` _outside_ the container. (`cat` is not important. Any command that does not exit immediately should work.)
4. Run `docker top c1 ae -o uid=PID -o command=COMMAND`

**Describe the results you received:**

    PID                 COMMAND
    1000                cat SECRET=foobar

**Describe the results you expected:**

Host-side process should not be shown in `docker top`

**Additional information you deem important (e.g. issue happens only occasionally):**

If an user is allowed to run `docker top`, he/she is likely allowed to run `docker run --privilege` as well and do anything for entire the host.
So basically the issue is not critical in typical deployment.

(But it is not 100%: he/she may be only allowed to run some specific docker commands via `sudo` or some kind of RPC proxy.)

### PR DESCRIPTION
**Pull request by AkihiroSuda**
**\- What I did**
Fix #24357

**\- How I did it**
Please look at `validatePSArgs`

**\- How to verify it**

```
$ docker top c1 ae -o uid=PID -o command=COMMAND
Error response from daemon: specifying uid=PID is not allowed
```

Signed-off-by: Akihiro Suda suda.akihiro@lab.ntt.co.jp

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thanks @akihirosuda

ping @justincormack @nathanmcauley PTAL

**Comment by thaJeztah** [comment]
ping @nathanmccauley PTAL

File: daemon/top_unix.go [hunk 1]
```diff
@@ -5,12 +5,25 @@ package daemon
...
+	"regexp"
 	"strconv"
 	"strings"
 
 	"github.com/docker/engine-api/types"
 )
 
+func validatePSArgs(psArgs string) error {
+	re := regexp.MustCompile("-o\\s*(.*)=\\s*PID")
+	group := re.FindStringSubmatch(psArgs)
```
Comment (justincormack) [inline daemon/top_unix.go]: This is not quite sufficient, as the header parsing is using `strings.Fields` and that parses based on unicode whitespace, so you can do `docker top d0b6b1e905ae ae -o uid= PID  -o command=COMMAND` with any unicode whitespace character between the `=` and `PID`. Changing the code to split strictly on ASCII spaces seems to be the best fix.

Comment (justincormack) [inline daemon/top_unix.go]: In fact you can relabel the columns in several ways using UTF8 whitespace eg `docker top 7a709f6ac6ae axe -o 'uid=UID PID' -o pid=COMMAND -o command=`

File: daemon/top_unix.go [hunk 2]
```diff
@@ -5,12 +5,25 @@ package daemon
...
 	"os/exec"
+	"regexp"
 	"strconv"
 	"strings"
 
 	"github.com/docker/engine-api/types"
 )
 
+func validatePSArgs(psArgs string) error {
+	re := regexp.MustCompile("-o\\s*(.*)=\\s*PID")
```
Comment (tonistiigi) [inline daemon/top_unix.go]: `-o` can also be `o`, `-O`, `O`, `--format`

File: daemon/top_unix.go (hunk 1 above)
Comment (AkihiroSuda) [inline daemon/top_unix.go]: Thank you, I'll update PR

**Comment by AkihiroSuda** [comment]
Thank you @justincormack @tonistiigi 
Updated PR.

Test case for unicode space (U+2003, E2 80 83):
```
$ docker top c1 ae -o uid= PID -o command=COMMAND
Error response from daemon: Couldn't find PID field in ps output
```

Test case for 'o':
```
$ docker top c1 aeo uid=PID -o command=COMMAND
Error response from daemon: specifying "uid=PID" is not allowed
```

**Comment by justincormack** [comment]
Thanks! LGTM.

**Comment by justincormack** [comment]
As this is a security fix I added to the 1.12 milestone.

File: daemon/top_unix.go [hunk 3]
```diff
@@ -5,49 +5,45 @@ package daemon
...
-
-	pids, err := daemon.containerd.GetPidsForContainer(container.ID)
-	if err != nil {
-		return nil, err
+func validatePSArgs(psArgs string) error {
+	// NOTE: \\s does not detect unicode whitespaces.
+	// So we use fieldsASCII instead of strings.Fields in parsePSOutput.
+	// See https://github.com/docker/docker/pull/24358
+	re := regexp.MustCompile("\\s+([^\\s]*)=\\s*PID")
+	group := re.FindStringSubmatch(psArgs)
```
Comment (tonistiigi) [inline daemon/top_unix.go]: Because this only matches one time it would not catch `-o pid=PID2 -o uid=PID`

Comment (AkihiroSuda) [inline daemon/top_unix.go]: thank you, updated PR

File: daemon/top_unix_test.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	tests := map[string]bool{
+		"ae -o uid=PID":             true,
+		"ae -o \"uid= PID\"":        true,  // ascii space (0x20)
+		"ae -o \"uid= PID\"":        false, // unicode space (U+2003, 0xe2 0x80 0x83)
+		"ae o uid=PID":              true,
+		"aeo uid=PID":               true,
+		"ae -O uid=PID":             true,
+		"ae -o pid=PID2 -o uid=PID": true,
+		"ae -o pid=PID":             false,
+		"ae -o pid=PID -o uid=PIDX": true, // FIXME: we do not need to prohibit this
```
Comment (justincormack) [inline daemon/top_unix_test.go]: As there are no default names that match this it seems unimportant for now.

**Comment by justincormack** [comment]
new version LGTM after @tonistiigi issue has been fixed.

**Comment by justincormack** [comment]
Test failures went away on rebuild, so green now.

**Comment by tonistiigi** [comment]
LGTM

@diogomonica do you want to verify this as well?

**Comment by diogomonica** [comment]
LGTM

**Comment by thaJeztah** [comment]
thanks @diogomonica. merging

**Pull request merged**